tests/graph/test_opt.py

```diff
@@ -691,3 +693,122 @@ def local_opt_2(fgraph, node):
             "optimizer: rewrite local_opt_2 replaces Op2(y, y) with [Op2.0]"
             in capres.out
         )
+
+
+def test_local_optimizer_str():
+    @local_optimizer([op1, MyOp])
+    def local_opt_1(fgraph, node):
+        pass
+
+    assert str(local_opt_1) == "local_opt_1"
+    res = repr(local_opt_1)
+    assert res.startswith("FromFunctionLocalOptimizer(")
+    assert "Op1" in res
+    assert "local_opt_1" in res
+
+
+def test_local_optimizer():
+
+    with pytest.raises(ValueError):
+
+        @local_optimizer([])
+        def local_bad_1(fgraph, node):
+            return node.outputs
+
+    with pytest.raises(TypeError):
+
+        @local_optimizer([None])
+        def local_bad_2(fgraph, node):
+            return node.outputs
+
+    x = MyVariable("x")
+    y = MyVariable("y")
+
+    o1 = op1(x, y)
+
+    class MyNewOp(MyOp):
+        pass
+
+    o2 = MyNewOp("MyNewOp")(x, y)
+
+    class MyNewOp2(MyOp):
+        pass
+
+    o3 = MyNewOp2("MyNewOp2")(x, y)
+
+    fgraph = FunctionGraph([x, y], [o1, o2, o3], clone=False)
+
+    hits = [0]
+
+    @local_optimizer([op1, MyNewOp])
+    def local_opt_1(fgraph, node, hits=hits):
+        hits[0] += 1
+        return node.outputs
+
+    # This is allowed by the `op1` in `tracks`
+    local_opt_1.transform(fgraph, fgraph.outputs[0].owner)
+    assert hits[0] == 1
+
+    # This is allowed by the `MyOp` in `tracks`
+    local_opt_1.transform(fgraph, fgraph.outputs[1].owner)
+    assert hits[0] == 2
+
+    # This is not allowed by `tracks`
+    local_opt_1.transform(fgraph, fgraph.outputs[2].owner)
+    assert hits[0] == 2
+
+
+def test_TrackingLocalOptimizer():
+    @local_optimizer(None)
+    def local_opt_1(fgraph, node):
+        pass
+
+    @local_optimizer([op1])
+    def local_opt_2(fgraph, node):
+        pass
+
+    @local_optimizer([Op])
+    def local_opt_3(fgraph, node):
+        pass
+
+    @local_optimizer([MyOp])
+    def local_opt_4(fgraph, node):
+        pass
+
+    @local_optimizer([MyOp])
+    def local_opt_5(fgraph, node):
+        pass
+
+    tracker = LocalOptTracker()
+    tracker.add_tracker(local_opt_1)
+    tracker.add_tracker(local_opt_2)
+    tracker.add_tracker(local_opt_3)
+    tracker.add_tracker(local_opt_4)
+    tracker.add_tracker(local_opt_5)
+
+    assert tracker.tracked_instances == {op1: [local_opt_2]}
+    assert tracker.tracked_types == {
+        Op: [local_opt_3],
+        MyOp: [local_opt_4, local_opt_5],
+    }
+    assert tracker.untracked_opts == [local_opt_1]
+
+    res = tracker.get_trackers(op1)
+    assert res == [local_opt_4, local_opt_5, local_opt_3, local_opt_2, local_opt_1]
+
+    class MyNewOp(Op):
+        def perform(self, *args):
+            pass
+
+    new_op = MyNewOp()
+
+    res = tracker.get_trackers(new_op)
+    assert res == [local_opt_3, local_opt_1]
```

Comment:
This is really neat! 

Any idea of how much slower the tracks pre-filtering should be now?

Reply:
Not much at all, especially since the MRO-based matching calls are memoized.